aragon/buidler-aragon
# Install apps from external networks (dependencies)

With very few extra lines of code fetch Aragon Apps from external networks (Mainnet), deploy them locally and consume them in the local dev environment

To achieve this, only requires this code added to the postDao hook in `test/projects/token-wrapper/buidler.config.ts`
```js
aragon: {
  hooks: {
    postDao: async ({ appInstaller }) => {
      const vault = await appInstaller("vault");
      const finance = await appInstaller("finance", {
        initializeArgs: [vault.address, 60 * 60 * 24 * 31]
      });

      await vault.createPermission("TRANSFER_ROLE", finance.address);
      await finance.createPermission("CREATE_PAYMENTS_ROLE");
    }
  }
}

await vault.createPermission("TRANSFER_ROLE", finance.address);
await finance.createPermission("CREATE_PAYMENTS_ROLE");
```

This allows installing any App that is available in publicly accessible networks. The API of is

https://github.com/aragon/buidler-aragon/blob/79e95df98876e35ca5789e81632069650a1c7d55/src/types.ts#L78-L81

https://github.com/aragon/buidler-aragon/blob/79e95df98876e35ca5789e81632069650a1c7d55/src/types.ts#L89-L94

https://github.com/aragon/buidler-aragon/blob/79e95df98876e35ca5789e81632069650a1c7d55/src/types.ts#L71-L76

## To test

Make sure your dependencies are in sync and execute the counter test project as normal

```
npm i
npm run dev

cd test/token-wrapper/counter
npx buidler start
```

You should be able to see the Finance app installed in the DAO at the latest version of mainnet

![Screenshot from 2020-03-18 02-34-12](https://user-images.githubusercontent.com/35266934/76916583-2b245680-68c1-11ea-95ef-3346bf628d08.png)

## Discussion / next steps

- What is the best hook to expose this functionality? `postDao`, `preInit`, `other`
- What name is more comprehensible? `appInstaller()`, `installApp()`, `other`
- Is it possible to generate a generic working deploy bytecode that would wrap the bytecode fetched from `eth.getCode()`?
- Cache the deploy bytecode of repo + version to prevent recurring fetches
- *Nice to have*: Cache the IPFS content locally so the dev can work offline
